pghoard/compressor.py

```diff
@@ -263,3 +281,74 @@ def set_state_defaults_for_site(self, site):
                     "count": 0
                 },
             }
+
+
+class WALFileDeleterThread(Thread):
+    """Deletes files which got compressed by the Compressor Thread, but keeps one file around
+
+    pg_receivewal, after some hickup, will use the files to compute the next wal segment to download.
+    if there are none, it will start at the current server position and so might miss wal segments. This
+    would mean that the backup is incomplete until the next base backup is taken and the service would
+    fail to be restored.
+
+    So the idea is to only unlink all files but the latest one.
+    """
+    def __init__(
+        self,
+        config: Dict,
+        wal_file_deletion_queue: Queue,
+        metrics: Metrics,
+    ):
+        super().__init__()
+        self.log = logging.getLogger("WALFileDeleter")
+        self.config = config
+        self.metrics = metrics
+        self.wal_file_deletion_queue = wal_file_deletion_queue
+        self.running = True
+        self.to_be_deleted_files: Dict[str, Set[str]] = {}
+        self.log.debug("WALFileDeleter initialized")
+
+    def run(self):
+        event: Optional[Dict] = None
+        while self.running:
+            if event is None:
+                try:
+                    event = self.wal_file_deletion_queue.get(timeout=self.config.get("deleter_event_get_timeout", 1))
+                except Empty:
+                    continue
+            try:
+                if event["type"] == "QUIT":
+                    break
+                if event["type"] == "delete_file":
+                    site = event["site"]
+                    local_path = event["local_path"]
+                    if site not in self.to_be_deleted_files:
+                        self.to_be_deleted_files[site] = set()
+                    self.to_be_deleted_files[site].add(local_path)
+                else:
+                    raise RuntimeError("Received bad event")
+                event = None
+                self.deleted_unneeded_files()
+            except Exception as ex:  # pylint: disable=broad-except
+                self.log.exception("Problem handling event %r: %s: %s", event, ex.__class__.__name__, ex)
+                self.metrics.unexpected_exception(ex, where="wal_file_deleter_run")
+                # If we have a problem, just keep running for now, at the worst we accumulate files
+                event = None
+                continue
+
+        self.log.debug("Quitting WALFileDeleter")
+
+    def deleted_unneeded_files(self):
+        """Deletes all but the latest file in the current list of files"""
+        for site, to_be_deleted_files_per_site in self.to_be_deleted_files.items():
+            if len(to_be_deleted_files_per_site) <= 1:
+                # Nothing to do (yet)
+                continue
```

Comment:
This is redundant, the `[:-1]` on the next line already covers that (including for empty set).

Reply:
I would like to keep this as "explicit is better than implicit"